Enhance pre-commit hooks to catch all flake8 errors that CI/CD checks

## Problem
The pre-commit hooks missed a flake8 E402 error that was caught by the CI/CD pipeline. This caused a PR (#26) to fail CI checks even though local pre-commit hooks passed.

## Specific Error
```
tests/integration/steps/error_logging_steps.py:15:1: E402 module level import not at top of file
```

## Root Cause
The pre-commit hook configuration may not be running flake8 on all the same files that the CI/CD pipeline checks, or it may have different configuration settings.

## Proposed Solution
1. Review and align pre-commit hook configuration with CI/CD checks
2. Ensure flake8 runs on all Python files in both src/ and tests/ directories
3. Verify that the same flake8 configuration is used in both environments
4. Consider adding a check that runs the exact same commands as CI/CD before allowing commits

## Acceptance Criteria
- [ ] Pre-commit hooks catch all errors that CI/CD would catch
- [ ] No more surprises where local checks pass but CI fails
- [ ] Documentation updated to explain the alignment between local and CI checks

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -21,12 +21,13 @@ repos:
     hooks:
       - id: black
         language_version: python3.11
+        args: ['src/', 'tests/']
 
   - repo: https://github.com/pycqa/flake8
     rev: 7.0.0
     hooks:
       - id: flake8
-        args: ['--max-line-length=88', '--extend-ignore=E203,W503']
+        args: ['--max-line-length=88', '--extend-ignore=E203,W503', 'src/', 'tests/']
         additional_dependencies: [flake8-docstrings]
 
   - repo: https://github.com/pre-commit/mirrors-mypy
@@ -49,3 +50,14 @@ repos:
     hooks:
       - id: ruff
         args: ['--fix']
+
+  - repo: local
+    hooks:
+      - id: ci-flake8-check
+        name: CI/CD Flake8 Check
+        entry: flake8
+        language: system
+        types: [python]
+        pass_filenames: false
+        args: ['src/', 'tests/']
+        description: Run the exact same flake8 command as CI/CD pipeline

--- a/CLAUDE.md
+++ b/CLAUDE.md
@@ -168,11 +168,18 @@ Beyond the universal TDD standards:
    - Run `pre-commit run --all-files` to see all issues
    - For persistent issues, fix manually or use `--no-verify` flag
 
-3. **Notion API Rate Limits**
+3. **CI/CD and Pre-commit Alignment**
+   - Pre-commit hooks are configured to match CI/CD checks exactly
+   - Both check `src/` and `tests/` directories with the same flake8 configuration
+   - Run `python scripts/validate_ci_alignment.py` to verify alignment
+   - If pre-commit passes, CI should also pass (no surprises!)
+   - The `ci-flake8-check` hook runs the exact same command as CI/CD
+
+4. **Notion API Rate Limits**
    - Default is 3 requests/second
    - Adjust `rate_limit_requests_per_second` in config if needed
 
-4. **Large Vault Processing**
+5. **Large Vault Processing**
    - Use batch_size configuration to control memory usage
    - Enable progress_bar to monitor progress
 

--- a/scripts/validate_ci_alignment.py
+++ b/scripts/validate_ci_alignment.py
@@ -1,0 +1,233 @@
+#!/usr/bin/env python3
+"""Validate that pre-commit hooks and CI/CD checks are aligned.
+
+This script ensures that local pre-commit hooks will catch all the same
+errors that the CI/CD pipeline would catch, preventing CI failures after
+local checks pass.
+"""
+
+import re
+import subprocess  # nosec B404
+import sys
+from pathlib import Path
+from typing import Dict, List, Tuple
+
+import yaml
+
+
+def extract_ci_commands(ci_file: Path) -> Dict[str, List[str]]:
+    """Extract linting commands from CI workflow file."""
+    with open(ci_file) as f:
+        ci_config = yaml.safe_load(f)
+
+    commands = {}
+    lint_job = ci_config.get("jobs", {}).get("lint", {})
+
+    for step in lint_job.get("steps", []):
+        if "run" in step and "name" in step:
+            name = step["name"]
+            if any(
+                tool in name.lower() for tool in ["black", "flake8", "mypy", "ruff"]
+            ):
+                # Extract the actual command
+                run_commands = step["run"].strip().split("\n")
+                for cmd in run_commands:
+                    cmd = cmd.strip()
+                    if cmd and not cmd.startswith("#"):
+                        commands[name] = cmd
+
+    return commands
+
+
+def extract_precommit_configs(precommit_file: Path) -> Dict[str, Dict]:
+    """Extract hook configurations from pre-commit config."""
+    with open(precommit_file) as f:
+        config = yaml.safe_load(f)
+
+    hooks = {}
+    for repo in config.get("repos", []):
+        for hook in repo.get("hooks", []):
+            hook_id = hook.get("id", "")
+            if hook_id in ["black", "flake8", "mypy", "ruff", "ci-flake8-check"]:
+                hooks[hook_id] = {
+                    "args": hook.get("args", []),
+                    "files": hook.get("files", ""),
+                    "pass_filenames": hook.get("pass_filenames", True),
+                }
+
+    return hooks
+
+
+def validate_flake8_alignment(
+    ci_cmd: str, precommit_hooks: Dict[str, Dict]
+) -> List[str]:
+    """Validate that flake8 configurations are aligned."""
+    issues = []
+
+    # Extract directories from CI command
+    ci_dirs = re.findall(r"flake8\s+(.+)", ci_cmd)
+    if ci_dirs:
+        ci_dirs = ci_dirs[0].split()
+
+    # Check both flake8 hooks
+    for hook_name in ["flake8", "ci-flake8-check"]:
+        if hook_name not in precommit_hooks:
+            continue
+
+        hook_config = precommit_hooks[hook_name]
+        hook_args = hook_config["args"]
+
+        # Check if directories are specified
+        precommit_dirs = [arg for arg in hook_args if arg in ["src/", "tests/"]]
+
+        if set(ci_dirs) != set(precommit_dirs):
+            issues.append(
+                f"{hook_name}: Directories mismatch - CI checks {ci_dirs}, "
+                f"pre-commit checks {precommit_dirs or 'default'}"
+            )
+
+    return issues
+
+
+def validate_black_alignment(
+    ci_cmd: str, precommit_hooks: Dict[str, Dict]
+) -> List[str]:
+    """Validate that black configurations are aligned."""
+    issues = []
+
+    if "black" not in precommit_hooks:
+        issues.append("black: Missing from pre-commit hooks")
+        return issues
+
+    # Extract options and directories from CI command
+    # ci_has_check = "--check" in ci_cmd  # Not used currently
+    ci_dirs = re.findall(r"black\s+(?:--check\s+)?(.+)", ci_cmd)
+    if ci_dirs:
+        ci_dirs = ci_dirs[0].split()
+
+    # Pre-commit black runs with --check by default
+    hook_args = precommit_hooks["black"]["args"]
+    precommit_dirs = [arg for arg in hook_args if not arg.startswith("-")]
+
+    if ci_dirs and not precommit_dirs:
+        issues.append(
+            f"black: CI checks specific directories {ci_dirs}, "
+            "but pre-commit doesn't specify directories"
+        )
+
+    return issues
+
+
+def run_test_comparison() -> Tuple[bool, List[str]]:
+    """Run both CI commands and pre-commit to compare results."""
+    print("\nRunning test comparison...")
+    comparison_issues = []
+
+    # Create a test file with known issues
+    test_file = Path("test_validation.py")
+    test_file.write_text(
+        '''
+import os
+import sys
+
+# This will trigger E402
+sys.path.append('.')
+
+import json  # E402: module level import not at top of file
+
+
+def badly_formatted_function( ):
+    """This function has formatting issues."""
+    x=1+2  # No spaces around operators
+    return x
+'''
+    )
+
+    try:
+        # Run flake8 as CI would
+        ci_result = subprocess.run(
+            ["flake8", "test_validation.py"], capture_output=True, text=True  # nosec
+        )
+        # ci_errors = (
+        #     ci_result.stdout.strip().split("\n") if ci_result.returncode != 0 else []
+        # )  # Not used currently
+
+        # Run pre-commit
+        pc_result = subprocess.run(
+            ["pre-commit", "run", "--files", "test_validation.py"],
+            capture_output=True,
+            text=True,  # nosec
+        )
+
+        # Check if pre-commit caught the E402 error
+        if "E402" in ci_result.stdout and "E402" not in pc_result.stdout:
+            comparison_issues.append(
+                "Pre-commit did not catch E402 error that CI would catch"
+            )
+
+        return len(comparison_issues) == 0, comparison_issues
+
+    finally:
+        test_file.unlink(missing_ok=True)
+
+
+def main():
+    """Main validation function."""
+    project_root = Path(__file__).parent.parent
+    ci_file = project_root / ".github" / "workflows" / "ci.yml"
+    precommit_file = project_root / ".pre-commit-config.yaml"
+
+    if not ci_file.exists():
+        print(f"Error: CI workflow file not found at {ci_file}")
+        sys.exit(1)
+
+    if not precommit_file.exists():
+        print(f"Error: Pre-commit config not found at {precommit_file}")
+        sys.exit(1)
+
+    print("Validating CI/CD and pre-commit alignment...")
+    print("=" * 60)
+
+    # Extract configurations
+    ci_commands = extract_ci_commands(ci_file)
+    precommit_hooks = extract_precommit_configs(precommit_file)
+
+    all_issues = []
+
+    # Validate each tool
+    for name, cmd in ci_commands.items():
+        print(f"\nChecking {name}...")
+        print(f"CI command: {cmd}")
+
+        if "flake8" in cmd.lower():
+            issues = validate_flake8_alignment(cmd, precommit_hooks)
+            all_issues.extend(issues)
+        elif "black" in cmd.lower():
+            issues = validate_black_alignment(cmd, precommit_hooks)
+            all_issues.extend(issues)
+
+        if not issues:
+            print("✓ Aligned with pre-commit")
+
+    # Run practical test
+    test_passed, test_issues = run_test_comparison()
+    all_issues.extend(test_issues)
+
+    # Report results
+    print("\n" + "=" * 60)
+    if all_issues:
+        print("❌ VALIDATION FAILED - Issues found:")
+        for issue in all_issues:
+            print(f"  - {issue}")
+        sys.exit(1)
+    else:
+        print("✅ VALIDATION PASSED - Pre-commit and CI/CD are aligned!")
+        print("\nRecommendations:")
+        print("  - Run 'pre-commit run --all-files' before pushing")
+        print("  - If pre-commit passes, CI should also pass")
+        print("  - Never add '# noqa' comments without careful consideration")
+        sys.exit(0)
+
+
+if __name__ == "__main__":
+    main()

--- a/tests/unit/test_readme_validation.py
+++ b/tests/unit/test_readme_validation.py
@@ -1,0 +1,151 @@
+"""Unit tests for README validation."""
+
+import re
+import unittest
+from pathlib import Path
+
+
+class TestReadmeValidation(unittest.TestCase):
+    """Test cases for validating README.md content."""
+
+    def setUp(self):
+        """Set up test fixtures."""
+        self.readme_path = Path("README.md")
+        if self.readme_path.exists():
+            self.readme_content = self.readme_path.read_text()
+            self.readme_lines = self.readme_content.splitlines()
+        else:
+            self.readme_content = ""
+            self.readme_lines = []
+
+    def test_readme_exists(self):
+        """Test that README.md exists in project root."""
+        self.assertTrue(
+            self.readme_path.exists(), "README.md not found in project root"
+        )
+
+    def test_no_bare_python_commands(self):
+        """Test that README doesn't contain bare 'python' commands."""
+        # Pattern to match 'python' followed by space, but not 'python3'
+        pattern = r"\bpython(?!3)\s"
+
+        bare_python_lines = []
+        for i, line in enumerate(self.readme_lines):
+            if re.search(pattern, line):
+                bare_python_lines.append((i + 1, line))
+
+        self.assertEqual(
+            len(bare_python_lines),
+            0,
+            f"Found bare 'python' commands on lines: {bare_python_lines}",
+        )
+
+    def test_python3_used_in_venv_setup(self):
+        """Test that virtual environment setup uses python3."""
+        venv_lines = [
+            (i + 1, line)
+            for i, line in enumerate(self.readme_lines)
+            if "venv" in line and "-m" in line
+        ]
+
+        for line_num, line in venv_lines:
+            self.assertIn(
+                "python3",
+                line,
+                f"Line {line_num} should use 'python3' for venv setup: {line}",
+            )
+
+    def test_python3_used_in_migrate_commands(self):
+        """Test that migrate.py commands use python3."""
+        migrate_lines = [
+            (i + 1, line)
+            for i, line in enumerate(self.readme_lines)
+            if "migrate.py" in line
+        ]
+
+        for line_num, line in migrate_lines:
+            if "python" in line:  # Only check lines that have python commands
+                self.assertIn(
+                    "python3",
+                    line,
+                    f"Line {line_num} should use 'python3' for migrate.py: {line}",
+                )
+
+    def test_all_code_blocks_use_python3(self):
+        """Test that all code blocks with Python commands use python3."""
+        in_code_block = False
+        code_block_lines = []
+
+        for i, line in enumerate(self.readme_lines):
+            if line.strip().startswith("```"):
+                in_code_block = not in_code_block
+            elif in_code_block and "python " in line:
+                code_block_lines.append((i + 1, line))
+
+        for line_num, line in code_block_lines:
+            # Check if it's a bare python command
+            if re.search(r"\bpython(?!3)\s", line):
+                self.fail(
+                    f"Line {line_num} in code block uses 'python' "
+                    f"instead of 'python3': {line}"
+                )
+
+    def test_consistency_across_readme(self):
+        """Test that Python command usage is consistent throughout README."""
+        # Count python3 vs python usage
+        python3_count = sum(1 for line in self.readme_lines if "python3" in line)
+        python_pattern = r"\bpython(?!3)\s"
+        python_count = sum(
+            1 for line in self.readme_lines if re.search(python_pattern, line)
+        )
+
+        # If we have any python3 commands, we shouldn't have bare python commands
+        if python3_count > 0:
+            self.assertEqual(
+                python_count,
+                0,
+                f"Inconsistent usage: found {python3_count} 'python3' "
+                f"and {python_count} 'python' commands",
+            )
+
+    def test_installation_section_uses_python3(self):
+        """Test that installation section specifically uses python3."""
+        in_installation = False
+        installation_lines = []
+
+        for i, line in enumerate(self.readme_lines):
+            if "## Installation" in line or "## Setup" in line:
+                in_installation = True
+            elif line.startswith("##") and in_installation:
+                in_installation = False
+            elif in_installation and "python" in line:
+                installation_lines.append((i + 1, line))
+
+        for line_num, line in installation_lines:
+            if re.search(r"\bpython(?!3)\s", line):
+                self.fail(
+                    f"Installation section line {line_num} should use 'python3': {line}"
+                )
+
+    def test_examples_section_uses_python3(self):
+        """Test that examples section uses python3."""
+        in_examples = False
+        example_lines = []
+
+        for i, line in enumerate(self.readme_lines):
+            if "## Usage" in line or "## Examples" in line or "## Example" in line:
+                in_examples = True
+            elif line.startswith("##") and in_examples:
+                in_examples = False
+            elif in_examples and "python" in line:
+                example_lines.append((i + 1, line))
+
+        for line_num, line in example_lines:
+            if re.search(r"\bpython(?!3)\s", line):
+                self.fail(
+                    f"Examples section line {line_num} should use 'python3': {line}"
+                )
+
+
+if __name__ == "__main__":
+    unittest.main()

